File: go/1_two_sum.go
package main

// Naive solution (Time: O(n^2) / Space: O(n^2))
func twoSumNaive(nums []int, target int) []int {
	for i := range nums {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return nil
}

// Optimal solution (Time: O(n) / Space: O(n))
func twoSumOptimal(nums []int, target int) []int {
	numMap := make(map[int]int)

	for i, num := range nums {
		complement := target - num
		if j, ok := numMap[complement]; ok {
			return []int{j, i}
		}
		numMap[num] = i
	}
	return nil
}

File: typescript/1_two_sum.ts
// Naive solution (Time: O(n^2) / Space: O(n^2))
function twoSumNaive(nums: number[], target: number): number[] {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  }
  return [];
}

// Optimal solution (Time: O(n) / Space: O(n))
function twoSumOptimal(nums: number[], target: number): number[] {
  const numMap = new Map<number, number>();

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    const complement = target - num;

    if (numMap.has(complement)) {
      return [numMap.get(complement)!, i];
    }
    numMap.set(num, i);
  }
  return [];
}
